torch/csrc/jit/codegen/cuda/lower_trivial_reductions.cpp

```diff
@@ -0,0 +1,103 @@
+#include <torch/csrc/jit/codegen/cuda/dispatch.h>
+#include <torch/csrc/jit/codegen/cuda/instrumentation.h>
+#include <torch/csrc/jit/codegen/cuda/ir_iostream.h>
+#include <torch/csrc/jit/codegen/cuda/ir_utils.h>
+#include <torch/csrc/jit/codegen/cuda/iter_visitor.h>
+#include <torch/csrc/jit/codegen/cuda/lower_trivial_reductions.h>
+#include <torch/csrc/jit/codegen/cuda/root_domain_map.h>
+
+#include <unordered_set>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+bool isDerivedFromTrivialReduction(TensorView* tv, IterDomain* id);
+
+bool traverseToRFactorTensor(TensorView* tv, IterDomain* root_id) {
+  TORCH_INTERNAL_ASSERT(
+      root_id->definition() == nullptr, "Not root IterDomain: ", root_id);
+
+  if (tv->definition() == nullptr) {
+    // This is an input tensor, so no rafactor tensor to traverse.
+    return false;
+  }
+
+  const auto& inputs = tv->definition()->inputs();
+
+  if (inputs.size() != 1 || !inputs[0]->isA<TensorView>() ||
+      tv->definition()->getExprType() != ExprType::ReductionOp) {
+    // No rfactor producer found
+    return false;
+  }
+
+  auto producer = inputs[0]->as<TensorView>();
+
+  if (!producer->hasRFactor()) {
+    return false;
+  }
+
+  auto c2p = PairwiseRootDomainMap(producer, tv)
+                 .mapConsumerToProducer(tv->domain(), producer->domain());
+
+  auto producer_id_it = c2p.find(root_id);
+  if (producer_id_it == c2p.end()) {
+    // No matching producer is found. Stop traversing.
+    return false;
+  }
+
+  auto producer_root_id = producer_id_it->second;
+
+  return isDerivedFromTrivialReduction(producer, producer_root_id);
+}
+
+bool isDerivedFromTrivialReduction(TensorView* tv, IterDomain* id) {
+  auto id_inputs = InputsOf::output(id->fusion(), id);
+  for (auto root_id : ir_utils::filterByType<IterDomain>(id_inputs)) {
+    if (root_id->isReduction() && root_id->rawExtent()->isOneInt()) {
+      continue;
+    }
+    // If not possible to prove the root ID is trivial, see if the ID
+    // is derived from a rfactor tensor and, if so, continue the
+    // analysis at the rfactor tensor.
+    if (!traverseToRFactorTensor(tv, root_id)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+} // namespace
+
+std::unordered_set<IterDomain*> detectTrivialReductionDerivedDomains(
+    Fusion* fusion) {
+  auto used_vals = DependencyCheck::getAllValsBetween(
+      {fusion->inputs().begin(), fusion->inputs().end()}, fusion->outputs());
+
+  std::unordered_set<IterDomain*> trivial_reductions;
+
+  for (auto tv : ir_utils::filterByType<TensorView>(used_vals)) {
+    for (auto id : tv->domain()->domain()) {
+      if (isDerivedFromTrivialReduction(tv, id)) {
```

Comment:
Seems this is because all vals between doesn't actually grab values explicitly in between provided values. Seems like we should really fix that, as it's really misleading that it doesn't. Shouldn't be hard to fix:
https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/iter_visitor.cpp#L377
Should only do that if there are no provided dependencies, and then if no provided dependencies we should insert the front of the stack at this point https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/iter_visitor.cpp#L372 to vals
similar to https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/iter_visitor.cpp#L503-L507

Reply:
Are you saying `grabAllValsBetween` grabs all vals in a dependency chain if no given dependency val appears in the chain? That is indeed counter-intuitive, and should be easy to fix.

Reply:
I'm saying it seems like it would make sense to me to do it in the reverse direction. Start from `tv3`'s root/rfactor domain, and propagate to leaves. Which could easily be done with `getAllValsBetween` if the behavior was fixed.

Reply:
If you think it makes more sense as is, I'm not 100% firm in this position, but the logic as is simply seems a bit convoluted to me.

Reply:
I'm sorry I meant 'tv2` not `tv3`. The root domains of `tv2` are:

```
tv2 root: iS11{i1} rS12{4}
tv2 maybe-rfactor root: iS11{i1} rS12{4}
```

How do we know `rS12{4}` is a trivial reduction if we start analysis from the root or maybe-rfactor root domains?

Reply:
My initial attempt was actually like as you suggested. It just looked at the root domain. 61f37db. 

However, I realized it doesn't work for the case like above.